src/cluster_asm.c

```diff
@@ -1669,23 +1673,38 @@ int slotRangesSnapshotSaveRio(int req, rio *rdb, int *error) {
                     robj key;
                     initStaticStringObject(key, kvobjGetKey(o));
 
-                    if (rioWriteBulkCount(rdb, '*', 5) == 0) goto werr;
-                    if (rioWriteBulkString(rdb, "RESTORE", 7) == 0) goto werr;
-                    if (rioWriteBulkObject(rdb, &key) == 0) goto werr;
-                    if (rioWriteBulkLongLong(rdb, expiretime == -1 ? 0 : expiretime) == 0) goto werr;
-
-                    /* Create the DUMP encoded representation. */
-                    rio payload;
-                    createDumpPayload(&payload, o, &key, i);
-                    sds buf = payload.io.buffer.ptr;
-                    if (rioWriteBulkString(rdb, buf, sdslen(buf)) == 0) {
+                    /* If non-string object that is not too big or module object that
+                     * not supporting aof_rewrite, use RESTORE to import data.
+                     * Generally RDB binary format is more efficient, but it may cause
+                     * block in the destination if the object is too large, so fall back
+                     * to AOF format if necessary. */
+                    if ((o->type != OBJ_STRING && getObjectLength(o) <= AOF_REWRITE_ITEMS_PER_CMD) ||
```

Comment:
getObjectLength() will always return zero for module type. Maybe we should exclude module types from this length check.